lib/Sema/CSDiagnostics.cpp

```diff
@@ -7311,6 +7311,19 @@ bool MissingContextualTypeForNil::diagnoseAsError() {
   return true;
 }
 
+bool CouldNotInferPlaceholderType::diagnoseAsError() {
+  // If this placeholder was explicitly written out by the user, they can maybe
+  // fix things by specifying an actual type.
+  if (auto *typeExpr = getAsExpr<TypeExpr>(getAnchor())) {
+    if (typeExpr->getLoc().isValid()) {
+      emitDiagnostic(diag::could_not_infer_placeholder);
```

Comment:
Do we have access to any potential bindings here? It would be very cool to surface notes that tell the user what we tried.

Reply:
Good idea! I'll give it a look at see if there's a good way to surface any of that information...

Reply:
Hmm, I would think that if a non-hole potential binding was attempted and didn't work, a more specific fix would be recorded. If several different bindings were attempted and none of them worked but the solutions all had the same score, we need to diagnose that through the ambiguity mechanism, similar to the "conflicting arguments to generic parameter" message. Here are some different examples with generic arguments, and I think placeholders should behave similarly:

```swift
protocol P {}

struct Generic<T>  {
  init() {}
  init(arg: T) where T: P {}
  init(arg1: T, arg2: T) {}
}

func test() {
  // no bindings for T -> missing generic arg fix 
  let generic1: Generic = .init() // error: Generic parameter 'T' could not be inferred

  // one binding for T that didn't work -> specific fix 
  let generic2: Generic = .init(arg: 1) // error: Initializer 'init(arg:)' requires that 'Int' conform to 'P'

  // several bindings for T and none worked -> ambiguity
  let generic3: Generic = .init(arg1: 1, arg2: "") // error: Conflicting arguments to generic parameter 'T' ('Int' vs. 'String')
}
```

One thing that would be cool, but is probably out of scope for this initial implementation, is recording any requirements on missing placeholders so the user has more information about what it needs to be filled in with.